inference-engine/thirdparty/clDNN/src/gpu/loop_gpu.cpp

```diff
@@ -0,0 +1,399 @@
+        int64_t execution_condition = read_scalar_value(initial_execution_mem);
+
+        // shortcut of current_iteration memory in body network (slice of input)
+        memory_impl* current_iteration_mem = nullptr;
+        if (node.is_current_iteration_used()) {
+            const primitive_id& current_iteration_id = node.get_current_iteration_id();
+            current_iteration_mem = &body_network->get_primitive(current_iteration_id)->output_memory();
+        }
+
+        // shortcut of execution_condition memory in body network
+        memory_impl* execution_condition_mem = nullptr;
+        if (node.is_execution_condition_used()) {
+            const primitive_id& condition_id = node.get_condition_id();
+            execution_condition_mem = &body_network->get_primitive(condition_id)->output_memory();
+        }
+
+        // output memory must be set before input_memory to set backedge memory properly
+        if (!instance.preproc_memories_done) {
+            preprocess_output_memory(instance);
+            preprocess_input_memory(instance);
+            preprocess_backedge_memory(instance);
+            instance.preproc_memories_done = true;
+        }
+
+        int64_t current_iteration = 0;
+        if (node.is_current_iteration_used()) {
+            write_scalar_value(*current_iteration_mem, current_iteration);
+        }
+
+        const auto& concatenated_input_mem_mappings = instance.concatenated_input_mem_mappings;
+        const auto& concatenated_output_mem_mappings = instance.concatenated_output_mem_mappings;
+        std::vector<event_impl::ptr> loop_carried_dep;
+        while (current_iteration < trip_count && execution_condition) {
+            // Copy & Set sliced input memory offset
+            for (size_t i = 0; i < instance.concatenated_input_mem_mappings.size(); ++i) {
+                const auto& concatenated_input = concatenated_input_mem_mappings.at(i);
+                memory_impl::ptr mem = concatenated_input.get_sliced_mem(static_cast<int>(current_iteration));
+                // set input mem
+                if (current_iteration == 0) {
+                    body_network->set_input_data(concatenated_input.sliced_data_id, *mem);
+                } else {
+                    concatenated_input.sliced_data_prim->set_output_memory(*mem);
+                }
+            }
+
+            // Set backedges
+            for (const auto& backedge_memory_mapping : instance.backedge_memory_mappings) {
+                backedge_memory_mapping.setup_iteration(current_iteration);
+            }
+
+            // Set sliced output memory
+            for (const auto& concat_output_mem_mapping : concatenated_output_mem_mappings) {
+                concat_output_mem_mapping.setup_concatenated_output_memory(current_iteration);
+            }
+            if (current_iteration == 0) {
+                body_network->execute(events);
+            } else {
+                body_network->execute(loop_carried_dep);
```

Comment:
I think we can remove this branching. 
std::vector<even_impl::ptr> loop_carried_dep = std::copy(events)...;

while (cur_iter < trip_count..) {
   ...
   body_network->execute(loop_carried_dep);
   loop_carried_dep.clear();
   for ( ..) {
        loop_carried_dep.emplace_back(body_event);
   }
}

Reply:
Applied